runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -1219,11 +1219,11 @@ private void setAndVerifyOutputTimestamp() {
       Instant windowExpiry = window.maxTimestamp().plus(allowedLateness);
       if (TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
         checkArgument(
-            !outputTimestamp.isAfter(target),
+            !outputTimestamp.isAfter(windowExpiry),
```

Comment:
Actually now that I think about it, the output timestamp really should be before the end of the window. That is a requirement on `WindowFn.assign` and other timestamp adjusting methods.

Reply:
Require the output timestamp to be before the end of the window is actually quite logical requirement, but is a breaking change. It will break code that sets timer (without output timestamp) to endOfWindow + lateness. Therefore I think we must keep the check as it is. Yes, it might result in data from a particular window being emitted for a different window (which really seems like a bug).

Reply:
GIving it one more thought, I think we should actually drop the Precondition for the output timestamp all together. Limiting it to any (future) timestamp seems inconsistent with the fact, that there is nothing stopping the user from using `outputWithTimestamp` and output arbitrarily future-bound value. The only thing the timer output timestamp actually influences is the output watermark. The output watermark is bound to be at most the value of input watermark, so setting a watermark hold to something higher is a no-op (semantically). Therefore this check should be removed completely.
@kennknowles @reuvenlax WDYT?